README.md
# mlr3learners.randomforest

<!-- badges: start -->
[![R CMD Check via {tic}](https://github.com/mlr3learners/mlr3learners.randomforest/workflows/R%20CMD%20Check%20via%20{tic}/badge.svg?branch=master)](https://github.com/mlr3learners/mlr3learners.randomforest/actions)
![Parameter Check](https://github.com/mlr3learners/mlr3learners.randomforest/workflows/Parameter%20Check/badge.svg?branch=master)
[![codecov](https://codecov.io/gh/mlr3learners/mlr3learners.randomforest/branch/master/graph/badge.svg)](https://codecov.io/gh/mlr3learners/mlr3learners.randomforest)
[![StackOverflow](https://img.shields.io/badge/stackoverflow-mlr3-orange.svg)](https://stackoverflow.com/questions/tagged/mlr3)
<!-- badges: end -->

Adds `randomforest()` from the {randomforest} package to {mlr3}.

Install the latest release of the package via

```r
install.packages("mlr3learners.randomforest")
```

by following the instructions in the [mlr3learners.drat README](https://github.com/mlr3learners/mlr3learners.drat).

Alternatively, you can install the latest version of {mlr3learners.randomforest} from Github with:

```r
remotes::install_github("mlr3learners/mlr3learners.randomforest")
```


NEWS.md
# mlr3learners.randomforest 0.1.2.9000

- Same as previous version.


# mlr3learners.randomforest 0.1.2

- More fixes for `importance()` method.


# mlr3learners.randomforest 0.1.1.9000

- Same as previous version.


# mlr3learners.randomforest 0.1.1

- Fix `importance()` calculation.


# mlr3learners.randomforest 0.1.0.9000

- Internal changes only.


# mlr3learners.randomforest 0.1.0

* Initial release


tests/testthat.R
library(testthat)
library(mlr3learners.randomforest)

test_check("mlr3")


tests/testthat/test_regr_randomForest.R
context("regr.randomForest")

test_that("autotest", {
  learner = LearnerRegrRandomForest$new()
  learner$param_set$values = list(ntree = 20, importance = "mse")
  expect_learner(learner)
  result = run_autotest(learner)
  expect_true(result, info = result$error)
})

test_that("Learner produces identical results as package version.", {
  set.seed(20191111)
  data(Boston, package = "MASS")
  Boston$chas = as.factor(Boston$chas)
  Boston$rad = as.factor(Boston$rad)
  task = TaskRegr$new(id = "Boston", backend = Boston, target = "crim")
  learner = LearnerRegrRandomForest$new()
  learner$param_set$values = list(ntree = 30L, importance = "mse")
  train_set = sample(task$nrow, 0.8 * task$nrow)
  test_set = setdiff(seq_len(task$nrow), train_set)
  set.seed(20191111)
  learner$train(task, row_ids = train_set)
  prediction = learner$predict(task, row_ids = test_set)

  set.seed(20191111)
  model = randomForest::randomForest(x = Boston[train_set, task$feature_names], y = Boston$crim[train_set], ntree = 30L, importance = TRUE)
  pred = predict(model, newdata = Boston[test_set, task$feature_names], type = "response")

  expect_true(all.equal(prediction$response, unname(pred)))
})


tests/testthat/test_classif_randomForest.R
context("classif.randomForest")

test_that("autotest", {
  learner = LearnerClassifRandomForest$new()
  learner$param_set$values = list(ntree = 20L, importance = "gini")
  expect_learner(learner)
  result = run_autotest(learner)
  expect_true(result, info = result$error)
})

test_that("Learner produces identical results as package version (multiclass, response).", {
  set.seed(20191111)
  data(Boston, package = "MASS")
  Boston$chas = as.factor(Boston$chas)
  Boston$rad = as.factor(Boston$rad)
  task = TaskClassif$new(id = "Boston", backend = Boston, target = "rad")
  learner = LearnerClassifRandomForest$new()
  learner$param_set$values = list(ntree = 30L, importance = "gini")
  train_set = sample(task$nrow, 0.8 * task$nrow)
  test_set = setdiff(seq_len(task$nrow), train_set)
  set.seed(20191111)
  learner$train(task, row_ids = train_set)
  prediction = learner$predict(task, row_ids = test_set)
  
  set.seed(20191111)
  model = randomForest::randomForest(x = Boston[train_set, task$feature_names], y = Boston$rad[train_set], ntree = 30L, importance = TRUE)
  pred = predict(model, newdata = Boston[test_set, task$feature_names], type = "response")
  
  expect_true(all.equal(prediction$response, as.factor(unname(pred))))
})

test_that("Learner produces identical results as package version (multiclass, prob).", {
  set.seed(20191111)
  data(Boston, package = "MASS")
  Boston$chas = as.factor(Boston$chas)
  Boston$rad = as.factor(Boston$rad)
  task = TaskClassif$new(id = "Boston", backend = Boston, target = "rad")
  learner = LearnerClassifRandomForest$new()
  learner$param_set$values = list(ntree = 30L, importance = "gini")
  learner$predict_type <- "prob"
  train_set = sample(task$nrow, 0.8 * task$nrow)
  test_set = setdiff(seq_len(task$nrow), train_set)
  set.seed(20191111)
  learner$train(task, row_ids = train_set)
  prediction = learner$predict(task, row_ids = test_set)
  
  set.seed(20191111)
  model = randomForest::randomForest(x = Boston[train_set, task$feature_names], y = Boston$rad[train_set], ntree = 30L, importance = TRUE)
  pred = predict(model, newdata = Boston[test_set, task$feature_names], type = "prob")
  
  expect_true(all.equal(unname(prediction$prob), unname(pred)))
})

test_that("Learner produces identical results as package version (twoclass, response).", {
  set.seed(20191111)
  data(Boston, package = "MASS")
  Boston$chas = as.factor(Boston$chas)
  Boston$rad = as.factor(Boston$rad)
  task = TaskClassif$new(id = "Boston", backend = Boston, target = "chas")
  learner = LearnerClassifRandomForest$new()
  learner$param_set$values = list(ntree = 30L, importance = "gini")
  train_set = sample(task$nrow, 0.8 * task$nrow)
  test_set = setdiff(seq_len(task$nrow), train_set)
  set.seed(20191111)
  learner$train(task, row_ids = train_set)
  prediction = learner$predict(task, row_ids = test_set)
  
  set.seed(20191111)
  model = randomForest::randomForest(x = Boston[train_set, task$feature_names], y = Boston$chas[train_set], ntree = 30L, importance = TRUE)
  pred = predict(model, newdata = Boston[test_set, task$feature_names], type = "response")
  
  expect_true(all.equal(prediction$response, as.factor(unname(pred))))
})

test_that("Learner produces identical results as package version (twoclass, prob).", {
  set.seed(20191111)
  data(Boston, package = "MASS")
  Boston$chas = as.factor(Boston$chas)
  Boston$rad = as.factor(Boston$rad)
  task = TaskClassif$new(id = "Boston", backend = Boston, target = "chas")
  learner = LearnerClassifRandomForest$new()
  learner$param_set$values = list(ntree = 30L, importance = "gini")
  learner$predict_type <- "prob"
  train_set = sample(task$nrow, 0.8 * task$nrow)
  test_set = setdiff(seq_len(task$nrow), train_set)
  set.seed(20191111)
  learner$train(task, row_ids = train_set)
  prediction = learner$predict(task, row_ids = test_set)
  
  set.seed(20191111)
  model = randomForest::randomForest(x = Boston[train_set, task$feature_names], y = Boston$chas[train_set], ntree = 30L, importance = TRUE)
  pred = predict(model, newdata = Boston[test_set, task$feature_names], type = "prob")
  
  expect_true(all.equal(unname(prediction$prob), unname(pred)))
})


R/zzz.R
#' @import data.table
#' @import paradox
#' @import mlr3misc
#' @importFrom R6 R6Class
#' @importFrom mlr3 mlr_learners LearnerClassif LearnerRegr
"_PACKAGE"

# nocov start
register_mlr3 = function(libname, pkgname) {
  # get mlr_learners dictionary from the mlr3 namespace
  x = utils::getFromNamespace("mlr_learners", ns = "mlr3")

  # add the learner to the dictionary
  x$add("classif.randomForest", LearnerClassifRandomForest)
  x$add("regr.randomForest", LearnerRegrRandomForest)
}

.onLoad = function(libname, pkgname) { # nolint
  register_mlr3()
  setHook(packageEvent("mlr3", "onLoad"), function(...) register_mlr3(),
    action = "append")
}

.onUnload = function(libpath) { # nolint
  event = packageEvent("mlr3", "onLoad")
  hooks = getHook(event)
  pkgname = vapply(hooks, function(x) environment(x)$pkgname, NA_character_)
  setHook(event, hooks[pkgname != "mlr3learners.randomforest"],
    action = "replace")
}
# nocov end


R/LearnerRegrRandomForest.R
#' @title Regression Random Forest Learner
#'
#' @name mlr_learners_regr.randomForest
#'
#' @description
#' Random forest learner.
#' \CRANpkg{randomForest} from package {randomForest}.
#'
#' @references
#' Breiman, L. (2001).
#' Random Forests
#' Machine Learning
#' \url{https://doi.org/10.1023/A:1010933404324}
#'
#' @export
LearnerRegrRandomForest = R6Class("LearnerRegrRandomForest",
  inherit = LearnerRegr,

  public = list(

    #' @description
    #' Creates a new instance of this [R6][R6::R6Class] class.
    initialize = function() {
      ps = ParamSet$new(
          params = list(
            ParamInt$new(
              id = "ntree", default = 500L, lower = 1L,
              tags = c("train", "predict")),
            ParamInt$new(id = "mtry", lower = 1L, tags = "train"),
            ParamLgl$new(id = "replace", default = TRUE, tags = "train"),
            ParamUty$new(id = "strata", tags = "train"),
            ParamUty$new(id = "sampsize", tags = "train"), # lower = 1L
            ParamInt$new(id = "nodesize", default = 5L, lower = 1L, tags = "train"),
            ParamInt$new(id = "maxnodes", lower = 1L, tags = "train"),
            ParamFct$new(
              id = "importance", default = FALSE,
              levels = c("mse", "nudepurity", "none", FALSE),
              special_vals = list(FALSE),
              tag = "train"),
            ParamLgl$new(id = "localImp", default = FALSE, tags = "train"),
            ParamLgl$new(id = "proximity", default = FALSE, tags = c("train", "predict")),
            ParamLgl$new(id = "oob.prox", tags = "train"),
            ParamLgl$new(id = "norm.votes", default = TRUE, tags = "train"),
            ParamLgl$new(id = "do.trace", default = FALSE, tags = "train"),
            ParamLgl$new(id = "keep.forest", default = TRUE, tags = "train"),
            ParamLgl$new(id = "keep.inbag", default = FALSE, tags = "train"),
            ParamLgl$new(id = "predict.all", default = FALSE, tags = "predict"),
            ParamLgl$new(id = "nodes", default = FALSE, tags = "predict")
          )
      )

      super$initialize(
        id = "regr.randomForest",
        packages = "randomForest",
        feature_types = c("integer", "numeric", "factor", "ordered"),
        predict_types = c("response"),
        param_set = ps,
        properties = c("weights", "importance", "oob_error"),
        man = "mlr3learners.randomforest::mlr_learners_regr.randomForest"
      )
    },

    #' @description
    #' The importance scores are extracted from the slot `importance`.
    #' Parameter 'importance' must be set to either `"mse"` or `"nodepurity"`.
    #' @return Named `numeric()`.
    importance = function() {
      if (is.null(self$model)) {
        stopf("No model stored")
      }
      imp = data.frame(self$model$importance)
      colnames(imp)[colnames(imp) == "X.IncMSE"] = "%IncMSE"
      ## correct for language error on special characters
      pars = self$param_set$get_values()

      scores = switch(pars[["importance"]],
        "mse" = imp[["%IncMSE"]],
        "nodepurity" = imp[["IncNodePurity"]],
        stop("No importance available. Try setting 'importance' to 'mse' or 'nodepurity'.")
      )

      sort(setNames(scores, rownames(imp)), decreasing = TRUE)
    },

    #' @description
    #' OOB errors are extracted from the model slot `mse`.
    #' @return `numeric(1)`.
    oob_error = function() {
      mean(self$model$mse)
    }
  ),

  private = list(

    .train = function(task) {

      pars = self$param_set$get_values(tags = "train")
      formula = task$formula()
      data = task$data()

      mlr3misc::invoke(randomForest::randomForest,
        formula = formula,
        data = data, .args = pars)
    },

    .predict = function(task) {

      pars = self$param_set$get_values(tags = "predict")
      newdata = task$data(cols = task$feature_names)
      type = self$predict_type

      pred = mlr3misc::invoke(predict, self$model,
        newdata = newdata,
        type = type, .args = pars)

      PredictionRegr$new(task = task, response = pred)
    }
  )
)


R/LearnerClassifRandomForest.R
#' @title Classification Random Forest Learner
#'
#' @name mlr_learners_classif.randomForest
#'
#' @description
#' Random forest learner.
#' \CRANpkg{randomForest} from package {randomForest}.
#'
#' @references
#' Breiman, L. (2001).
#' Random Forests
#' Machine Learning
#' \url{https://doi.org/10.1023/A:1010933404324}
#'
#' @export
LearnerClassifRandomForest = R6Class("LearnerClassifRandomForest",
  inherit = LearnerClassif,

  public = list(

    #' @description
    #' Creates a new instance of this [R6][R6::R6Class] class.
    initialize = function() {
      ps = ParamSet$new(
        params = list(
          ParamInt$new(
            id = "ntree", default = 500L, lower = 1L,
            tags = c("train", "predict")),
          ParamInt$new(id = "mtry", lower = 1L, tags = "train"),
          ParamLgl$new(id = "replace", default = TRUE, tags = "train"),
          ParamUty$new(id = "classwt", default = NULL, tags = "train"),
          ParamUty$new(id = "cutoff", tags = "train"),
          ParamUty$new(id = "strata", tags = "train"),
          ParamUty$new(id = "sampsize", tags = "train"),
          ParamInt$new(
            id = "nodesize", default = 1L, lower = 1L,
            tags = "train"),
          ParamInt$new(id = "maxnodes", lower = 1L, tags = "train"),
          ParamFct$new(
            id = "importance", default = FALSE,
            levels = c("accuracy", "gini", "none", FALSE),
            special_vals = list(FALSE),
            tag = "train"),
          ParamLgl$new(id = "localImp", default = FALSE, tags = "train"),
          ParamLgl$new(id = "proximity", default = FALSE, tags = c("train", "predict")),
          ParamLgl$new(id = "oob.prox", tags = "train"),
          ParamLgl$new(id = "norm.votes", default = TRUE, tags = "train"),
          ParamLgl$new(id = "do.trace", default = FALSE, tags = "train"),
          ParamLgl$new(id = "keep.forest", default = TRUE, tags = "train"),
          ParamLgl$new(id = "keep.inbag", default = FALSE, tags = "train"),
          ParamLgl$new(id = "predict.all", default = FALSE, tags = "predict"),
          ParamLgl$new(id = "nodes", default = FALSE, tags = "predict")
        )
      )

      super$initialize(
        id = "classif.randomForest",
        packages = "randomForest",
        feature_types = c("numeric", "factor", "ordered"),
        predict_types = c("response", "prob"),
        param_set = ps,
        properties = c("weights", "twoclass", "multiclass", "importance", "oob_error"),
        man = "mlr3learners.randomforest::mlr_learners_classif.randomForest"
      )
    },


    #' @description
    #' The importance scores are extracted from the slot `importance`.
    #' Parameter 'importance' must be set to either `"accuracy"` or `"gini"`.
    #' @return Named `numeric()`.
    importance = function() {
      if (is.null(self$model)) {
        stopf("No model stored")
      }
      imp = data.frame(self$model$importance)
      pars = self$param_set$get_values()

      scores = switch(pars[["importance"]],
        "accuracy" = imp[["MeanDecreaseAccuracy"]],
        "gini"     = imp[["MeanDecreaseGini"]],
        stop("No importance available. Try setting 'importance' to 'accuracy' or 'gini'.")
      )

      sort(setNames(scores, rownames(imp)), decreasing = TRUE)
    },

    #' @description
    #' OOB errors are extracted from the model slot `err.rate`.
    #' @return `numeric(1)`.
    oob_error = function() {
      mean(self$model$err.rate[, 1L])
    }
  ),

  private = list(
    .train = function(task) {

      pars = self$param_set$get_values(tags = "train")

      formula = task$formula()
      data = task$data()
      levs = levels(data[[task$target_names]])
      n_levels = length(levs)

      if (!"cutoff" %in% names(pars)) {
        cutoff = rep(1 / n_levels, n_levels)
      }
      if ("classwt" %in% names(pars)) {
        classwt = pars[["classwt"]]
        if (is.numeric(classwt) && length(classwt) == n_levels &&
          is.null(names(classwt))) {
          names(classwt) = levs
        }
      } else {
        classwt = NULL
      }
      if (is.numeric(cutoff) && length(cutoff) == n_levels &&
        is.null(names(cutoff))) {
        names(cutoff) = levs
      }
      mlr3misc::invoke(randomForest::randomForest,
        formula = formula,
        data = data, classwt = classwt, cutoff = cutoff, .args = pars)
    },

    .predict = function(task) {
      pars = self$param_set$get_values(tags = "predict")
      newdata = task$data(cols = task$feature_names)
      type = ifelse(self$predict_type == "response", "response", "prob")

      pred = mlr3misc::invoke(predict, self$model,
        newdata = newdata,
        type = type, .args = pars)

      if (self$predict_type == "response") {
        PredictionClassif$new(task = task, response = pred)
      } else {
        PredictionClassif$new(task = task, prob = pred)
      }
    }
  )
)
